### PR TITLE
Update S4070 to accept negative numbers

### LINKED ISSUE #7991: Fix Sxxxx FP/FN: Issue title
### Description

Get the S4070 on flaged enum 

### Repro steps

	[Flags]
	public enum Failing
	{
		None = 0,
                MyTest = 1 << 31
        }

### Expected behavior

No S4070

### Actual behavior

Get the S4070 on flaged enum 

### Related information

* Visual Studio version

### PR DESCRIPTION
**Pull request by costin-zaharia-sonarsource**
if their absolute value is a power of two.

Fixes #7991

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.CSharp/Rules/DoNotMarkEnumsWithFlags.cs [hunk 1]
```diff
@@ -67,35 +68,49 @@ public sealed class DoNotMarkEnumsWithFlags : SonarDiagnosticAnalyzer
...
+                      long x => (BigInteger)x,
+                      ulong x => (BigInteger)x,
+                      // this is a safeguard for the case when the enum is defined with a base type that is not currently supported by the compiler
+                      var x when BigInteger.TryParse(x.ToString(), out var value) => value,
+                      _ => null
+                  }
                 : null;
 
-        private static bool IsValidFlagValue(ulong? enumValue, List<ulong> allValues) =>
+        private static bool IsValidFlagValue(BigInteger? enumValue, IEnumerable<BigInteger> allValues) =>
```
Comment (sebastien-marichal) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/DoNotMarkEnumsWithFlags.cs:89]: Educational: Why did you change the type from `List` to `IEnumerable`?
Isn't there a risk of multiple enumeration if we drop the `ToList` on `allValues` line 58?

**Review by sebastien-marichal — APPROVED** [review]
LGTM!
I left one educational question.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/DoNotMarkEnumsWithFlags.cs (hunk 1 above)
Comment (costin-zaharia-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/DoNotMarkEnumsWithFlags.cs:89]: I think that the method should not be aware of how it is used. If it does the iteration only once and doesn't use API specific to the collection it should use the most generic type. This in some scenarios will allow lazy evaluation, or no evaluation at all if the path is not reached.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/DoNotMarkEnumsWithFlags.cs [hunk 2]
```diff
@@ -67,35 +68,49 @@ public sealed class DoNotMarkEnumsWithFlags : SonarDiagnosticAnalyzer
...
             enumValue.HasValue && (IsZeroOrPowerOfTwo(enumValue.Value) || IsCombinationOfOtherValues(enumValue.Value, allValues));
 
-        // See https://stackoverflow.com/questions/600293/how-to-check-if-a-number-is-a-power-of-2
-        private static bool IsZeroOrPowerOfTwo(ulong value) =>
-            (value & (value - 1)) == 0;
+        private static bool IsZeroOrPowerOfTwo(BigInteger value) =>
+            value.IsZero
+            || (value.Sign == -1
+                    ? BigInteger.Multiply(value, -1).IsPowerOfTwo
+                    : value.IsPowerOfTwo);
```
Comment (antonioaversa) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/DoNotMarkEnumsWithFlags.cs:96]: BigInteger.Abs(value).IsPowerOfTwo
```suggestion
|| BigInteger.Abs(value).IsPowerOfTwo
```

**Review by antonioaversa — APPROVED** [review]
LGTM, we should be able to use `BigInteger.Abs`.

**Comment by antonioaversa** [comment]
@costin-zaharia-sonarsource The safeguard and the `_ => null` seem difficult to cover.
I wonder whether we could just remove the safeguard, and return `null` whenever the type doesn't match one of the 8 types above.
I also wonder when `enumMember.HasConstantValue` is `false` and whether we can write an incomplete tuple to cover also `tuple => tuple.Member.DeclaringSyntaxReferences.FirstOrDefault()?.GetSyntax().GetLocation()`.

**Comment by costin-zaharia-sonarsource** [comment]
@antonioaversa I removed the safeguard. Regarding the tuple, I think that's a scenario that could not actually happen, to have an enum declaration with a member that doesn't have a location. It looks like defensive coding.
